KAFKA-14562 [1/2]: Implement epoch bump after every transaction

jolshan opened the pull request:
Implement server side changes for epoch bump but keep EndTxn as an unstable API until the client side changes are implemented. EndTxnResponse will return the producer ID and epoch for the transaction. Introduces new tagged fields to the TransactionLogValue to persist the clientTransactionVersion, previousProducerId, and nextProducerId to the log so that the state can be reloaded. See [KIP-890](https://cwiki.apache.org/confluence/display/KAFKA/KIP-890%3A+Transactions+Server-Side+Defense) for more details.

Small updates to naming of lastProducerId -> PreviousProducerId. Also cleans up the many TransactionMetadata constructors.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jolshan committed:
Start

jolshan committed:
Updates

jolshan committed:
Merge branch 'trunk' of https://github.com/apache/kafka into kafka-14562

jolshan committed:
Fix validEpochCheck

jolshan committed:
Update to include new field in transaction log and correctly check the right state for retries

jolshan committed:
Bug fix

jolshan committed:
Tests and fixes

jolshan committed:
Cleanups

jolshan commented:
I may go back and clean up some of the "magic" values in tests. I use 0 for TV 0 and 2 for TV 2, but I could use a named variable for these versions, or latest if I just want the latest.

jolshan committed:
Add transaction metadata transitions correctly.

jolshan committed:
Fix completing transactions

jolshan commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -517,9 +526,19 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
+            producerIdCopy = txnMetadata.producerId
+            producerEpochCopy = txnMetadata.producerEpoch
+            val currentTxnMetadataIsAtLeastTransactionsV2 = txnMetadata.clientTransactionVersion >= 2
+            if (txnMetadata.producerId != producerId && !(txnMetadata.previousProducerId == producerId && txnMetadata.producerEpoch == 0 && currentTxnMetadataIsAtLeastTransactionsV2))
```
Going to refactor this so it can also be used on line 638

jolshan commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -540,24 +572,29 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
+                  case Right(nextProducerId) =>
+                    Right(coordinatorEpoch, txnMetadata.prepareAbortOrCommit(nextState, clientTransactionVersion, nextProducerId, time.milliseconds()))
+                }
               case CompleteCommit =>
```
Could refactor this with isInvalidTxnTransition -- if not at least explain the retry case is handled above for v2 so it will always be invalid state transition

artemlivshits replied:
Yeah, I wonder if it would be more readable to keep this code as it was, but when we check producerId and epoch above just do a version-specific producer id and epoch check.

jolshan replied:
Originally I put the retry check above the epoch check since I thought it would be easier to do that first and not touch the check for the same epoch. The tricky thing is specifically that if we hit this code we have the same epoch which is wrong. (In other words, we should have the same OR a different epoch depending on the state -- but I can factor that all in the epoch check, it just means we will check the state in two places.)

jolshan commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -487,15 +519,22 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
   private def validProducerEpoch(transitMetadata: TxnTransitMetadata): Boolean = {
-    val transitEpoch = transitMetadata.producerEpoch
-    val transitProducerId = transitMetadata.producerId
+    val isAtLeastTransactionsV2 = transitMetadata.clientTransactionVersion >= 2
```
Want to refactor or at least make this code clearer

artemlivshits commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -181,6 +183,7 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
               newMetadata.producerEpoch,
               TransactionResult.ABORT,
               isFromClient = false,
+              clientTransactionVersion = 0,
```
Is using version=0 the right thing here?  I'd expect that we want epoch bump on aborted transactions.

jolshan replied:
In the fencing case, we have separate logic for bumping the epoch. It does something slightly different and I didn't want to change that logic in this PR.

jolshan replied:
I was thinking about this some more. We can support setting 0 here and only using the new epoch bump logic in the non prepareProducerFence/fromClient cases. But I also wonder, since these are not even initiated by the client, does the naming of this value even make sense? I do see that maybe we would want to use the tagged fields and extra features, so I will probably just use the server-side TV. But I am wondering if this name will be confusing.

artemlivshits commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -668,11 +709,23 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
+  }
+
+  def isRetryEndTxn(txnMetadata: TransactionMetadata, producerId: Long, producerEpoch: Short): Boolean = {
+    // The previous producer ID matches and the epoch is either + 1 the request epoch or 0 if the epoch overflowed.
```
Can we add some comments on how we check the conditions?  Also, I think if we structure the code to check retry conditions for (PrepareCommit, PrepareAbort) and then for (CompleteAbort, CompleteCommit), I think it may be more readable, especially if we annotate with comments for each case that explain the transitions.

jolshan replied:
I think if we want to redo the epoch check with the state transitions we can cover this. My only concern is if that ordering messes with the old handling. I will look into that

artemlivshits commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -594,7 +635,7 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
                       if (txnMetadata.producerId != producerId)
                         Left(Errors.INVALID_PRODUCER_ID_MAPPING)
-                      else if (txnMetadata.producerEpoch != producerEpoch)
+                      else if (txnMetadata.producerEpoch != producerEpoch && (requestIsAtLeastTransactionsV2 && txnMetadata.producerEpoch != producerEpoch + 1))
```
Should it be `txnMetadata.producerEpoch != producerEpoch && (!requestIsAtLeastTransactionsV2 || txnMetadata.producerEpoch != producerEpoch + 1)`?  As written we effectively ignore epoch check when `requestIsAtLeastTransactionsV2==false`.

jolshan replied:
I think I got my logic backwards here. I will take another look since as is the parentheses don't make sense.

artemlivshits commented on core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala:
```diff
@@ -79,7 +79,7 @@ object TransactionLog {
     // which enables flexible fields in records.
     val version: Short =
-      if (usesFlexibleRecords) 1 else 0
+      if (transactionVersionLevel > 1) 1 else 0
```
According to comments the condition should be `transactionVersionLevel >= 1`.

jolshan committed:
Refactor the transitions logic
Also return concurrent transactions if retry occurs in prepare states rather than none

artemlivshits commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -517,47 +526,90 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
+                  }
 
-                Right(coordinatorEpoch, txnMetadata.prepareAbortOrCommit(nextState, time.milliseconds()))
+                  nextProducerIdOrErrors match {
```
```
  nextProducerIdOrErrors.flatMap {
    nextProducerId =>
      Right(coordinatorEpoch, txnMetadata.prepareAbortOrCommit(nextState, clientTransactionVersion, nextProducerId, time.milliseconds()))
  }

artemlivshits commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -517,47 +526,90 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
+                }
               case CompleteCommit =>
-                if (txnMarkerResult == TransactionResult.COMMIT)
+                if (txnMarkerResult == TransactionResult.COMMIT && validCompleteTransition)
```
It looks like `validCompleteTransition` is effectively epoch check and should result in producer fenced error, but it results in invalid transition error.

jolshan replied:
Hmmm. I don't know if I agree. The producer is not fenced in this case. In this case, the epoch equals the txnMetadata and it is invalid. 

If the epoch is < the txn state, it will get caught in the fenced case above. I tried to explain in the comment, but maybe it is unclear.

artemlivshits replied:
This is not consistent with what we had before: unexpected epoch led to producer fenced error:
```
if ((isFromClient && producerEpoch != txnMetadata.producerEpoch) || producerEpoch < txnMetadata.producerEpoch)
              Left(Errors.PRODUCER_FENCED)
```
If we decide to redefine this condition, we should redefine it consistently for all states.

jolshan replied:
I don't think that's quite true because the epoch is expected when it is the same epoch -- it was just the wrong state. We just never had a case where same epoch == invalidState unless it was the wrong type of marker. 

I don't feel super strongly though since the result will be the same from the client side.

jolshan replied:
I have updated with the producer fenced approach

artemlivshits commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -517,47 +526,90 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
-            if (txnMetadata.producerId != producerId)
+            producerIdCopy = txnMetadata.producerId
+            producerEpochCopy = txnMetadata.producerEpoch
+            val currentTxnMetadataIsAtLeastTransactionsV2 = txnMetadata.clientTransactionVersion >= 2
```
I think the logic is mostly correct here (except maybe returning invalid state error where we'd normally return producer fenced error), I wonder if it could be more readable if we had a block of code that calculate 2 Booleans: isValidProducerId and isValidEpoch, each calculation would take the states and versions into consideration.

jolshan replied:
I think that is closer to what I had before -- though I focused on retries as that was what I thought was a non-trivial special case. 

I changed it to the current state since I thought we were trying to keep the current format where we have the various states enumerated.

artemlivshits replied:
I don't have a strong opinion, if you think this code structure is best, I think it works (I left a comment on the other issue separately).

jolshan committed:
Return PRODUCER_FENCED on all epoch errors

artemlivshits reviewed (approved):
LGTM

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -502,12 +507,16 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
                              producerEpoch: Short,
                              txnMarkerResult: TransactionResult,
                              isFromClient: Boolean,
+                             clientTransactionVersion: Short,
```
At first, I was a bit puzzle by this version here because it was not clear to me whether it was the version of the RPC or the TV. In order to make it explicit, I wonder if we should pass the `TransactionVersion` here. If possible, I would also advice to `clientTransactionVersion >= 2` condition within a method in `TransactionVersion`. We did this in the `GroupVersion` and I find it pretty elegant. What do you think?

jolshan replied:
The one thing that is tricky here is we also need to encode that the client sent the newest request (or in other words -- the client + the server support the new version and the client acknowledges that)

jolshan replied:
We discussed offline and I realized he was talkinga bout the type. I have also added the specified method.

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala:
```diff
@@ -79,7 +79,7 @@ object TransactionLog {
     // which enables flexible fields in records.
     val version: Short =
-      if (usesFlexibleRecords) 1 else 0
+      if (transactionVersionLevel >= 1) 1 else 0
```
If we pass `TransactionVersion` to this method too, we could have a `recordVersion` which returns the version to use here.

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -163,70 +163,64 @@ private[transaction] case object PrepareEpochFence extends TransactionState {
   * @param topicPartitions       current set of partitions that are part of this transaction
   * @param txnStartTimestamp     time the transaction was started, i.e., when first partition is added
   * @param txnLastUpdateTimestamp   updated when any operation updates the TransactionMetadata. To be used for expiration
+  * @param clientTransactionVersion    TransactionVersion used by the client when the staet was transitioned
```
nit: Should we re-align all the descriptions?

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -335,18 +329,32 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
-    prepareTransitionTo(newState, producerId, producerEpoch, lastProducerEpoch, txnTimeoutMs, topicPartitions.toSet,
-      txnStartTimestamp, updateTimestamp)
+  def prepareAbortOrCommit(newState: TransactionState, clientTransactionVersion: Short, nextProducerId: Long, updateTimestamp: Long): TxnTransitMetadata = {
+
```
nit: I suppose that we could remove this empty line.

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -335,18 +329,32 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
-      txnStartTimestamp, updateTimestamp)
+    val (updatedProducerId, updatedProducerEpoch) = if (clientTransactionVersion >= 2 && nextProducerId != RecordBatch.NO_PRODUCER_ID) {
+      // If we overflowed on epoch bump, we have to set it as the producer ID now the marker has been written.
+        (nextProducerId, 0.toShort)
+      } else {
+        (producerId, producerEpoch)
+      }
```
nit: The indentation seems be to off here.

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -335,18 +329,32 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
     hasFailedEpochFence = false
-    prepareTransitionTo(newState, producerId, producerEpoch, lastProducerEpoch, txnTimeoutMs, Set.empty[TopicPartition],
-      txnStartTimestamp, updateTimestamp)
+    val (updatedProducerId, updatedProducerEpoch) = if (clientTransactionVersion >= 2 && nextProducerId != RecordBatch.NO_PRODUCER_ID) {
```
Echoing on one of my previous comment, it would be great to have `clientTransactionVersion >= 2` centralized in one place.

dajac commented on transaction-coordinator/src/main/resources/common/message/TransactionLogValue.json:
```diff
@@ -37,6 +41,8 @@
     { "name": "TransactionStartTimestampMs", "type": "int64", "versions": "0+",
-      "about": "Time the transaction was started"}
+      "about": "Time the transaction was started"},
+    { "name": "clientTransactionVersion", "type": "int16", "default": 0, "taggedVersions": "1+", "tag": 2,
```
nit: `ClientTransactionVersion`?

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -668,11 +726,19 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
+  // When a client and server support V2, every endTransaction call bumps the producer epoch. When checking epoch, we want to
+  // check epoch + 1. Epoch bumps from PrepareEpochFence state are handled separately, so this method should not be used to check that case.
+  // Returns true if the transaction state epoch is the specified producer epoch + 1 and epoch bump on every transaction is expected.
+  def endTxnEpochBumped(txnMetadata: TransactionMetadata, producerEpoch: Short): Boolean = {
```
nit: Could we keep it private?

dajac reviewed (commented):
@jolshan Thanks for the patch. I left some comments for consideration.

jolshan committed:
Pass TransactionVersion around

jolshan committed:
Merge branch 'trunk' of https://github.com/apache/kafka into kafka-14562

dajac reviewed (approved):
LGTM. Thanks for the patch!

dajac commented:
Ah, I just noticed that they are failed tests that seem related. Could you please take a look into those?

jolshan committed:
Fix method

chia7712 commented on core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala:
```diff
@@ -78,17 +79,16 @@ object TransactionLog {
-    val version: Short =
-      if (usesFlexibleRecords) 1 else 0
-    MessageUtil.toVersionPrefixedBytes(version,
+    MessageUtil.toVersionPrefixedBytes(transactionVersionLevel.transactionLogValueVersion(),
```
Apologies for raising a question on this PR.


I noticed that when the TV is downgraded, the broker returns an `UnsupportedVersionException` to the producer, which seems to put the producer into a fatal error state.

```
    org.apache.kafka.common.KafkaException: Cannot execute transactional method because we are in an error state
        at app//org.apache.kafka.clients.producer.internals.TransactionManager.maybeFailWithError(TransactionManager.java:1173)
        at app//org.apache.kafka.clients.producer.internals.TransactionManager.lambda$beginAbort$3(TransactionManager.java:366)
        at app//org.apache.kafka.clients.producer.internals.TransactionManager.handleCachedTransactionRequestResult(TransactionManager.java:1282)
        at app//org.apache.kafka.clients.producer.internals.TransactionManager.beginAbort(TransactionManager.java:364)
        at app//org.apache.kafka.clients.producer.KafkaProducer.abortTransaction(KafkaProducer.java:895)
        at app//org.apache.kafka.clients.TransactionsDowngradeTest.test(TransactionsDowngradeTest.java:78)

        Caused by:
        org.apache.kafka.common.errors.UnsupportedVersionException: The version of API is not supported.
```
Is the exception above expected?

KIP-890 mentions "If we see a lower MV, we should abort the transaction and start again with the old protocol."

However, I'm not sure if "start again" implies that the producer application needs to be restarted (due to the fatal error), or if the producer client should be able to handle this gracefully

jolshan replied:
I think the idea was that the upgrade path would be handled gracefully, but the downgrade path was not guaranteed to do so. Just to clarify though, this unsupported version should only show up if the image itself is downgraded, not the MV/TV alone. 

That may be slightly different for downgrading from tv1 to tv0, but my understanding is that should also be only on image version downgrade. The record format downgrades are based on the semantics in https://cwiki.apache.org/confluence/display/KAFKA/KIP-915%3A+Txn+and+Group+Coordinator+Downgrade+Foundation. Can look into the case more specifically if we need.

chia7712 replied:
@jolshan Thanks for the detailed explanation and for sharing your insights!
